interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,588 @@
+            *[section.length for section in self.Section_list],
+            interpolation='linear')
+
+        # Reshaping to match the dimensions of the synchronous data program
+        self.section_length = self.section_length.reshape(
+            self.n_sections, self.cycle_time, self.use_turns)
+
+        # Getting the circumference and radius (including potential orbit
+        # bumps)
+        self.circumference = np.sum(self.section_length, axis=0)
+        self.radius = self.circumference / (2 * np.pi)
+
+        # Computing the revolution period on the design orbit
+        # as well as revolution frequency and angular frequency
+        self.t_rev_design = np.dot(self.section_length_design,
+                                   1 / (self.beta * c))
+        self.f_rev_design = 1 / self.t_rev_design
+        self.omega_rev_design = 2 * np.pi * self.f_rev_design
+
+        # Computing the time of flight in each section
+        # and the revolution period on the beam orbit including
+        # possible orbit bumps
+        self.t_rev = np.sum(np.array(
+            self.section_length / (self.beta * c)), axis=0)
+        self.f_rev = 1 / self.t_rev
+        self.omega_rev = 2 * np.pi * self.f_rev
+
+        # Recalculating the delta_E
+        self.delta_E = np.diff(self.energy, axis=1)
+        if self.n_turns > len(self.use_turns):
+            self.delta_E = np.zeros(self.energy.shape)
+            self._recalc_delta_E()
+
+        # Determining the momentum compaction orders defined in all sections
+        # The orders 1 and 2 are presently set by default to zeros if
+        # not defined in sections for the calculation of all orders of eta
+        self.alpha_orders = []
+        for section in self.Section_list:
+            self.alpha_orders += section.alpha_orders
+
+        # Add alpha orders if eta_orders is defined
+        self.eta_orders = kwargs.pop('eta_orders', 0)
+        if self.eta_orders >= 1:
+            self.alpha_orders += [1]
+        if self.eta_orders >= 2:
+            self.alpha_orders += [2]
+        if self.eta_orders >= 3:
+            warn_message = 'The eta_orders can only be computed up to eta_2!'
+            warnings.warn(warn_message)
+
+        self.alpha_orders = np.unique(self.alpha_orders)
+
+        # Getting the momentum compaction programs in all sections
+        # and combining the programs (the missing programs are filled
+        # with zeros).
+        # The programs are reshaped to the size of the momentum program
+        for order in self.alpha_orders:
+
+            alpha_prog = []
+            alpha_name = 'alpha_%d' % (order)
+
+            for section in self.Section_list:
+
+                if hasattr(section, alpha_name):
+                    alpha_prog.append(getattr(section, alpha_name))
+                else:
+                    alpha_prog.append(ring_programs.momentum_compaction(
+                        0, order=order))
```

Comment:
"Easier to ask for forgiveness than permission"

```suggestion
                try:
                    alpha_prog.append(getattr(section, alpha_name))
                except AttributeError:
                    alpha_prog.append(ring_programs.momentum_compaction(
                        0, order=order))
```

`hasattr` effectively does a `getattr` and returns `True` if there's no `AttributeError` and `False` if there is an `AttributeError`.  May as well do it ourselves.

Reply:
Side note:
It's faster to use `try/except` if there is no exception raised, but faster to use `if/else` if the condition is `False`.  Obviously this is a single call so it's not really relevant, but if something is run repeatedly and the `exception` is expected to be rare, it's faster to use `try/except` than `if/else`.

Reply:
Corrected

Thanks for the side note I was about to ask that specifically!